### PR TITLE
Submit embedded groups applications automatically

### PR DESCRIPTION
**Pull request by Johaney-s**
* After approval of VO application, applications to initially selected embedded groups are submitted

### CODE REVIEW

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java [hunk 1]
```diff
@@ -1567,6 +1568,98 @@ public Application approveApplication(PerunSession sess, int appId) throws Perun
...
+
+				try {
+					getFormForGroup(group);
+				} catch (FormNotExistsException e) {
+					createApplicationFormInGroup(sess, group);
+				}
+
+				submitApplication(sess, groupApplication, new ArrayList<>());
+			} catch (PerunException e) {
+				errorMessages = errorMessages.concat(System.lineSeparator() + e.getMessage());
```
Comment (Vojtech-Sassmann) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java:1654]: We should also add ids of the failed groups. I think that the best solution would be to add an extra property to the `EmbeddedGroupApplicationSubmitionError`. 
* `failedGroups` - this would be a map from `Integer` to `String`. For each failed group, it would store its id and appropriate message from the exception.

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java [hunk 2]
```diff
@@ -1567,6 +1568,98 @@ public Application approveApplication(PerunSession sess, int appId) throws Perun
...
+			}
+		}
+
+		if (errorMessages != "") {
```
Comment (Vojtech-Sassmann) [inline perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java:1658]: Strings should be compared only by using the `equals` method. Also, if you want to test, if a string is empty, there is the `.isEmpty()` method on all Strings.

File: perun-registrar-lib/src/test/java/cz/metacentrum/perun/registrar/RegistrarBaseIntegrationTest.java [hunk 3]
```diff
@@ -560,6 +564,60 @@ private static void applyForMembershipInVO(RegistrarManager registrarManager, Pe
 		registrarManager.createApplication(user, application, data);
 	}
 
+	@Test
+	public void testEmbeddedGroupsSubmission() throws PerunException {
+		GroupsManager groupsManager = perun.getGroupsManager();
+		ApplicationForm form = registrarManager.getFormForVo(vo);
+		form.setAutomaticApproval(false);
+		form.setAutomaticApprovalEmbedded(false);
```
Comment (Vojtech-Sassmann) [inline perun-registrar-lib/src/test/java/cz/metacentrum/perun/registrar/RegistrarBaseIntegrationTest.java:572]: I think this is not actually doing anything, since these changes are not saved to the DB.

**Review by Vojtech-Sassmann — CHANGES_REQUESTED** [review]

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java [hunk 4]
```diff
@@ -1567,6 +1568,98 @@ public Application approveApplication(PerunSession sess, int appId) throws Perun
...
+				groupApplication.setUser(app.getUser());
+				groupApplication.setType(AppType.EMBEDDED);
+				groupApplication.setVo(app.getVo());
+				groupApplication.setGroup(group);
+				groupApplication.setFedInfo(app.getFedInfo());
+				groupApplication.setExtSourceName(app.getExtSourceName());
+				groupApplication.setExtSourceType(app.getExtSourceType());
+				groupApplication.setExtSourceLoa(app.getExtSourceLoa());
+				groupApplication.setCreatedBy("Automatically generated");
+				groupApplication.setCreatedAt(new SimpleDateFormat("yyyy/MM/dd HH:mm:ss").format(new Date()));
```
[the remark this answers is not in the document]
Comment (Johaney-s) [reply perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java]: removed also from test

File: perun-registrar-lib/src/test/java/cz/metacentrum/perun/registrar/RegistrarBaseIntegrationTest.java (hunk 3 above)
Comment (Johaney-s) [reply perun-registrar-lib/src/test/java/cz/metacentrum/perun/registrar/RegistrarBaseIntegrationTest.java]: removed

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java (hunk 2 above)
Comment (Johaney-s) [reply perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java]: noted and changed to Map

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java (hunk 1 above)
Comment (Johaney-s) [reply perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java]: added

File: perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java [hunk 5]
```diff
@@ -1567,6 +1568,98 @@ public Application approveApplication(PerunSession sess, int appId) throws Perun
...
+
+				try {
+					getFormForGroup(group);
+				} catch (FormNotExistsException e) {
+					createApplicationFormInGroup(sess, group);
+				}
+
+				submitApplication(sess, groupApplication, new ArrayList<>());
+			} catch (PerunException e) {
```
[the remark this answers is not in the document]
Comment (Johaney-s) [reply perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java]: done

**Review by Vojtech-Sassmann — APPROVED** [review]
